File: Cpp/Array/704_Binary_Search.cpp
#include <vector>

using namespace std;

class Solution_Brute {
public:
    int search(vector<int>& nums, int target) {
        for (size_t i = 0; i < nums.size(); i++) {
            if (nums[i] == target) {
                return i;
            }
        }
        return -1;
    }
};

class Solution {
public:
    int search(vector<int>& nums, int target) {
        int left = 0;
        int right = nums.size() - 1;

        while (left <= right) {
            int middle = left + (right - left) / 2;

            if (nums[middle] == target) {
                return middle;
            }
            if (nums[middle] < target) {
                left = middle + 1;
            } else {
                right = middle - 1;
            }
        }

        return -1;
    }
};

int main() {
    Solution sol;
    vector<int> nums = {-1, 0, 3, 5, 9, 12};
    int result = sol.search(nums, 9);
    (void)result;
    return 0;
}

File: Python/Array/704 Binary Search.py
# Input: nums = [-1,0,3,5,9,12], target = 9
# Output: 4

# Reference Solution - Linear Search
class Solution_Brute:
    def search(self, nums: list[int], target: int) -> int:
        for index, value in enumerate(nums):
            if value == target:
                return index
        return -1


# Optimal Solution - Binary Search
class Solution:
    def search(self, nums: list[int], target: int) -> int:
        left = 0
        right = len(nums) - 1

        while left <= right:
            middle = left + (right - left) // 2

            if nums[middle] == target:
                return middle
            if nums[middle] < target:
                left = middle + 1
            else:
                right = middle - 1

        return -1


# Local Testing
nums = [-1, 0, 3, 5, 9, 12]
sol = Solution()
print(sol.search(nums, 9))
